pkg/k8sutils/k8sutils.go

```diff
@@ -0,0 +1,310 @@
+type Instance struct {
+	kubeClient kubernetes.Interface
+	k8sOps     k8s.Ops
+}
+
+// New returns a new instance of k8sutils
+func New(kubeconfig string) (*Instance, error) {
+	var cfg *rest.Config
+	var err error
+
+	if len(kubeconfig) == 0 {
+		kubeconfig = os.Getenv("KUBECONFIG")
+	}
+
+	if len(kubeconfig) > 0 {
+		logrus.Debugf("using kubeconfig: %s to create k8s client", kubeconfig)
+		cfg, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+	} else {
+		logrus.Debugf("will use in-cluster config to create k8s client", kubeconfig)
+		cfg, err = rest.InClusterConfig()
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("Error building kubeconfig: %s", err.Error())
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Instance{
+		kubeClient: kubeClient,
+		k8sOps:     k8s.Instance(),
+	}, nil
+}
+
+// GetPXSharedApps returns all deployments and statefulsets using Portworx storage class for shared volumes
+func (k *Instance) GetPXSharedApps() ([]apps_api.Deployment, []apps_api.StatefulSet, error) {
+	scs, err := k.GetPXSharedSCs()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var sharedDeps []apps_api.Deployment
+	var sharedSS []apps_api.StatefulSet
+	for _, sc := range scs {
+		deps, err := k.k8sOps.GetDeploymentsUsingStorageClass(sc)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		for _, d := range deps {
+			sharedDeps = append(sharedDeps, d)
+		}
+
+		ss, err := k.k8sOps.GetStatefulSetsUsingStorageClass(sc)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		for _, s := range ss {
+			sharedSS = append(sharedSS, s)
+		}
+	}
+
+	return sharedDeps, sharedSS, nil
+}
+
+// GetPXSharedSCs returns all storage classes that have the shared parameter set to true
+func (k *Instance) GetPXSharedSCs() ([]string, error) {
+	scs, err := k.kubeClient.Storage().StorageClasses().List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	var retList []string
+	for _, sc := range scs.Items {
+		params, err := k.k8sOps.GetStorageClassParams(&sc)
+		if err != nil {
+			return nil, err
+		}
+
+		if isShared, present := params["shared"]; present && isShared == "true" {
+			retList = append(retList, sc.Name)
+		}
+
+	}
+
+	return retList, nil
+}
+
+// ScaleSharedAppsToZero scales down PX shared apps to 0 replicas
+func (k *Instance) ScaleSharedAppsToZero() error {
+	sharedDeps, sharedSS, err := k.GetPXSharedApps()
+	if err != nil {
+		return err
+	}
+
+	logrus.Infof("found %d deployments and %d statefulsets to scale down", len(sharedDeps), len(sharedSS))
+
+	var valZero int32
+	for _, d := range sharedDeps {
+		deploymentName := d.Name
+		deploymentNamespace := d.Namespace
+		logrus.Infof("scaling down deployment: [%s] %s", deploymentNamespace, deploymentName)
+
+		t := func() (interface{}, bool, error) {
+			dCopy, err := k.kubeClient.Apps().Deployments(deploymentNamespace).Get(deploymentName, metav1.GetOptions{})
+			if err != nil {
+				if IsNotFoundErr(err) {
```

Comment:
use kubernetes/apimachinery/pkg/api/errors".IsNotFound  instead:

https://github.com/kubernetes/apimachinery/blob/master/pkg/api/errors/errors.go#L418

Reply:
fixed in incremental